src/OpenFeature/EventExecutor.cs

```diff
@@ -0,0 +1,368 @@
+        }
+
+        private void StartListeningAndShutdownOld(FeatureProviderReference newProvider, FeatureProviderReference oldProvider)
+        {
+            // check if the provider is already active - if not, we need to start listening for its emitted events
+            if (!this.IsProviderActive(newProvider))
+            {
+                this._activeSubscriptions.Add(newProvider);
+                var featureProviderEventProcessing = new Thread(this.ProcessFeatureProviderEventsAsync);
+                featureProviderEventProcessing.Start(newProvider);
+            }
+
+            if (oldProvider != null && !this.IsProviderBound(oldProvider))
+            {
+                this._activeSubscriptions.Remove(oldProvider);
+                var channel = oldProvider.Provider.GetEventChannel();
+                if (channel != null)
+                {
+                    channel.Writer.WriteAsync(new ShutdownSignal());
+                }
+            }
+        }
+
+        private bool IsProviderBound(FeatureProviderReference provider)
+        {
+            if (this._defaultProvider == provider)
+            {
+                return true;
+            }
+            foreach (var providerReference in this._namedProviderReferences.Values)
+            {
+                if (providerReference == provider)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private bool IsProviderActive(FeatureProviderReference providerRef)
+        {
+            return this._activeSubscriptions.Contains(providerRef);
+        }
+
+        private void EmitOnRegistration(FeatureProviderReference provider, ProviderEventTypes eventType, EventHandlerDelegate handler)
+        {
+            if (provider == null)
+            {
+                return;
+            }
+            var status = provider.Provider.GetStatus();
+
+            var message = "";
+            if (status == ProviderStatus.Ready && eventType == ProviderEventTypes.ProviderReady)
+            {
+                message = "Provider is ready";
+            }
+            else if (status == ProviderStatus.Error && eventType == ProviderEventTypes.ProviderError)
+            {
+                message = "Provider is in error state";
+            }
+            else if (status == ProviderStatus.Stale && eventType == ProviderEventTypes.ProviderStale)
+            {
+                message = "Provider is in stale state";
+            }
+
+            if (message != "")
+            {
+                try
+                {
+                    handler.Invoke(new ProviderEventPayload
+                    {
+                        ProviderName = provider.Provider.GetMetadata().Name,
+                        Type = eventType,
+                        Message = message
+                    });
+                }
+                catch (Exception exc)
+                {
+                    Debug.WriteLine(exc);
```

Comment:
```suggestion
                    Debug.WriteLine("Error running handler: " + exc.toString());
```

I think this additional context is helpful.

Reply:
I think also we should make this a `error` level log. Otherwise it will be tricky to trace errors in event handlers. We do this in [JS](https://github.com/open-feature/js-sdk/blob/072d2c7570274e003e32cd50807ed8a0839e7b4c/packages/shared/src/open-feature.ts#L94) and [Java](https://github.com/open-feature/java-sdk/blob/9851d4ba1c238b72673932badce04a4d1bc882fb/src/main/java/dev/openfeature/sdk/EventSupport.java#L142C36-L143C22)

But interested in what @kinyoklion @beeme1mr @Kavindu-Dodan think.

Reply:
I actually think there is a bit more of a problem than that. We are theoretically using `Microsoft.Extensions.Logging` and allowing customization of that. But it looks like that is only implemented at the client level, and API level doesn't have any logger instance.

So, it should be using a logger instance that can be set by a developer. Then once it is doing that, then likely it should be error level.

(`Debug.WriteLine` is a system diagnostics feature)

Reply:
Oh yes, good catch. I missed that entirely.

Reply:
alright, I will adapt it to use a `Microsoft.Extensions.Logging` logger

Reply:
I added a `SetLogger()` to ehe `Api` method that allows to set a `Microsoft.Extensions.Logging` implementation, which will propagated through to the `EventExecutor` (which is the only component of the `Api` that potentially logs something right now, but we may use this for other parts of the `Api` as well in the future).